### PR TITLE
Trampoline payments use per-channel fee and cltv

### PR DESCRIPTION
**Pull request by thomash-acinq**
Make sure trampoline users pay at least as much as the other users.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala [hunk 1]
```diff
@@ -95,10 +96,11 @@ object Graph {
                         pathsToFind: Int,
                         wr: Option[WeightRatios],
                         currentBlockHeight: Long,
-                        boundaries: RichWeight => Boolean): Seq[WeightedPath] = {
+                        boundaries: RichWeight => Boolean,
+                        isRelay: Boolean): Seq[WeightedPath] = {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala:100]: I would rename it to something that makes what it does more clear, for example `includeLocalChannel(Cost|Weight)` or `countLocalChannel(Cost|Weight)` (or on the contrary, if you find it more clear, that parameter could be for *ignoring* local channel costs: `ignoreLocalChannel(Cost|Weight)`)

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 2]
```diff
@@ -432,7 +432,7 @@ object Router {
 
   case class MultiPartParams(minPartAmount: MilliSatoshi, maxParts: Int)
 
-  case class RouteParams(randomize: Boolean, maxFeeBase: MilliSatoshi, maxFeePct: Double, routeMaxLength: Int, routeMaxCltv: CltvExpiryDelta, ratios: Option[WeightRatios], mpp: MultiPartParams) {
+  case class RouteParams(randomize: Boolean, maxFeeBase: MilliSatoshi, maxFeePct: Double, routeMaxLength: Int, routeMaxCltv: CltvExpiryDelta, ratios: Option[WeightRatios], mpp: MultiPartParams, isRelay: Boolean) {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala:435]: Same here, I would change the name of `isRelay` to something that more clearly explains that it governs whether we'll count local channel fees or not.

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala [hunk 3]
```diff
@@ -120,12 +120,13 @@ object NodeRelay {
 
   /** Compute route params that honor our fee and cltv requirements. */
   def computeRouteParams(nodeParams: NodeParams, amountIn: MilliSatoshi, expiryIn: CltvExpiry, amountOut: MilliSatoshi, expiryOut: CltvExpiry): RouteParams = {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala:122]: NB: we still verify in `validateRelay` above that the fee is at least our current default fee, so we ignore the fact that we may have channels that require lower fees and could relay this payment.

But I think it's the right behavior to have, trampoline is more computationally expensive than standard channel-relay so it makes sense to have a higher threshold than our cheapest channels.

File: eclair-core/src/test/scala/fr/acinq/eclair/router/RouteCalculationSpec.scala [hunk 4]
```diff
@@ -1662,6 +1662,14 @@ class RouteCalculationSpec extends AnyFunSuite with ParallelTestExecution {
     val fees = routes.map(_.fee)
     assert(fees.forall(_ == fees.head))
   }
+
+  test("can't relay if fee is not sufficient") {
+    val g = DirectedGraph(List(
+      makeEdge(1L, a, b, 1000 msat, 7000),
+    ))
+
+    val Failure(_) = findRoute(g, a, b, 10000000 msat, 10000 msat, numRoutes = 3, routeParams = DEFAULT_ROUTE_PARAMS.copy(isRelay = true), currentBlockHeight = 400000)
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/router/RouteCalculationSpec.scala:1671]: nit: we can be a little more precise:
```
assert(findRoute(g, a, b, 10000000 msat, 10000 msat, numRoutes = 3, routeParams = DEFAULT_ROUTE_PARAMS.copy(isRelay = true), currentBlockHeight = 400000) === Failure(RouteNotFound)
```

And I would also add the positive case for completeness:

```
assert(findRoute(g, a, b, 10000000 msat, 100000 msat, numRoutes = 3, routeParams = DEFAULT_ROUTE_PARAMS.copy(isRelay = true), currentBlockHeight = 400000).isSuccess)
```

**Review by t-bast — COMMENTED** [review]
Thanks, this is indeed simple enough to include asap ;)

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala (hunk 1 above)
Comment (thomash-acinq) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: Renamed to includeLocalChannelCost, I agree it's a better name.

**Review by t-bast — APPROVED** [review]
LGTM!

Make sure to choose the "Squash and merge" option when merging, and provide a good commit title and message.

**Pull request merged**